tests/h/util/session_tracker_test.py

```diff
@@ -0,0 +1,88 @@
+from __future__ import unicode_literals
+
+from uuid import uuid4
+import pytest
+
+from h.db.types import _get_urlsafe_from_hex
+from h.models import Annotation, Document
+from h.util.session_tracker import Tracker, ObjectState
+
+
+def generate_ann_id():
+    """
+    Generate a random annotation identifier in the encoded form used by the API.
+    """
+    return _get_urlsafe_from_hex(str(uuid4()))
+
+
+class TestTracker(object):
+
+    def test_uncommitted_changes_returns_unflushed_changes(self, tracker, session, expected_changes):
+        added_entry, changed_entry, deleted_entry = expected_changes
+
+        changes = tracker.uncommitted_changes()
+
+        assert added_entry in changes
+        assert changed_entry in changes
+        assert deleted_entry in changes
+
+    def test_uncommitted_changes_returns_flushed_changes(self, tracker, session, expected_changes):
+        added_entry, changed_entry, deleted_entry = expected_changes
+
+        session.flush()
+        changes = tracker.uncommitted_changes()
+
+        assert added_entry in changes
+        assert changed_entry in changes
+        assert deleted_entry in changes
+
+    def test_uncommitted_changes_does_not_return_committed_changes(self, tracker, session):
+        session.commit()
+        assert tracker.uncommitted_changes() == []
+
+    def test_uncommitted_changes_does_not_return_rolled_back_changes(self, tracker, session):
+        session.rollback()
+        assert tracker.uncommitted_changes() == []
+
+    @pytest.fixture
+    def expected_changes(self, changed_ann_id, deleted_ann_id):
+        added_entry = (Annotation, None, ObjectState.ADDED)
+
+        # FIXME - Figure out why the PK is a tuple rather than a primitive
+        changed_entry = (Annotation, (changed_ann_id,), ObjectState.CHANGED)
+        deleted_entry = (Annotation, (deleted_ann_id,), ObjectState.DELETED)
+
+        return (added_entry, changed_entry, deleted_entry)
+
+    @pytest.fixture
+    def changed_ann_id(self):
+        return generate_ann_id()
+
+    @pytest.fixture
+    def deleted_ann_id(self):
+        return generate_ann_id()
+
+    @pytest.fixture
+    def session(self, db_session, changed_ann_id, deleted_ann_id):
```

Comment:
I’m finding the use of the single mono-fixture tricky to follow: it makes it more difficult for me to see the intent of any individual test. Do you think there would be a way to break out individual test fixtures so that they’re a little more self-contained, or would that result in an unreasonable amount of duplication?